Realm/RLMSyncSubscription.mm

```diff
@@ -0,0 +1,483 @@
+    if ([errorMessage length] == 0) {
+        return NULL;
+    }
+    return [[NSError alloc]initWithDomain:RLMSyncErrorDomain
+                                               code:0
+                                           userInfo:@{NSLocalizedDescriptionKey : errorMessage}];
+}
+
+- (RLMSyncSubscriptionState)state {
+    switch (_subscriptionSet->state()) {
+        case realm::sync::SubscriptionSet::State::Uncommitted:
+        case realm::sync::SubscriptionSet::State::Pending:
+        case realm::sync::SubscriptionSet::State::Bootstrapping:
+            return RLMSyncSubscriptionStatePending;
+        case realm::sync::SubscriptionSet::State::Complete:
+            return RLMSyncSubscriptionStateComplete;
+        case realm::sync::SubscriptionSet::State::Error:
+            return RLMSyncSubscriptionStateError;
+        case realm::sync::SubscriptionSet::State::Superceded:
+            return RLMSyncSubscriptionStateSuperceded;
+    }
+}
+
+#pragma mark - Batch Update subscriptions
+
+- (void)write:(__attribute__((noescape)) void(^)(void))block {
+    return [self write:block onComplete:^(NSError**){}];
+}
+
+- (void)write:(__attribute__((noescape)) void(^)(void))block
+   onComplete:(void(^)(NSError**))completionBlock {
+    [self secureWrite];
+    _mutableSubscriptionSet = std::make_unique<realm::sync::MutableSubscriptionSet>(_subscriptionSet->make_mutable_copy());
+    self->isInWriteTransaction = true;
+    block();
+    try {
+        _subscriptionSet = std::make_unique<realm::sync::SubscriptionSet>(std::move(*_mutableSubscriptionSet).commit());
+        _mutableSubscriptionSet = nullptr;
+        self->isInWriteTransaction = false;
+    }
+    catch (std::exception error) {
+        NSError *err = [[NSError alloc] initWithDomain:@"subscription_set" code:-1 userInfo:@{@"reason":@(error.what())}];
+        completionBlock(&err);
+    }
+    _subscriptionSet->get_state_change_notification(realm::sync::SubscriptionSet::State::Complete)
+        .get_async([completionBlock](realm::StatusWith<realm::sync::SubscriptionSet::State> state) mutable noexcept {
+            if (state.is_ok()) {
+                auto value = state.get_value();
+                completionBlock(nil);
+            } else {
+                NSError* error = [[NSError alloc] initWithDomain:@"sync_subscriptions" code:state.get_status().code() userInfo:@{@"reason": @(state.get_status().reason().c_str())}];
+                completionBlock(&error);
+            }
+        });
+}
+
+typedef void(^RLMSyncSubscriptionCallback)(NSError * _Nullable error);
+
+- (BOOL)writeAsync:(__attribute__((noescape)) void(^)(void))block
+          callback:(RLMSyncSubscriptionCallback)callback {
+    [NSException raise:@"NotImplemented" format:@"Needs Implementation"];
+    return NULL;
+}
+
+#pragma mark - Find subscription
+
+- (nullable RLMSyncSubscription *)subscriptionWithName:(NSString *)name {
+    std::string str = std::string([name UTF8String]);
```

Comment:
Converting this to a std::string rather than passing `name.UTF8String` directly to `find()` is just a pessimization.